test_rmw_implementation/test/test_wait_set.cpp

```diff
@@ -0,0 +1,322 @@
+    var_name->internal_var_name ## s = static_cast<void **>(allocator.allocate( \
+        sizeof(void *) * size, allocator.state)); \
+    if (!var_name->internal_var_name ## s) { \
+      allocator.deallocate(static_cast<void *>(var_name), allocator.state); \
+      var_name->internal_var_name ## s = nullptr; \
+      break; \
+    } \
+    memset(var_name->internal_var_name ## s, 0, sizeof(void *) * size); \
+} while (0);
+
+// Macro to free memory of rmw_wait input variables
+#define FREE_MEMORY(var_name, internal_var_name) \
+  if (var_name->internal_var_name ## s) { \
+    allocator.deallocate(static_cast<void **>(var_name->internal_var_name ## s), allocator.state); \
+    var_name->internal_var_name ## s = nullptr; \
+    var_name->internal_var_name ## _count = 0; \
+  } \
+  if (var_name) { \
+    allocator.deallocate(static_cast<void *>(var_name), allocator.state); \
+    var_name = nullptr; \
+  }
+
+TEST_F(CLASSNAME(TestWaitSet, RMW_IMPLEMENTATION), rmw_create_wait_set)
+{
+  // Created a valid wait_set
+  rmw_wait_set_t * wait_set = rmw_create_wait_set(&context, 0);
+  EXPECT_NE(wait_set, nullptr) << rcutils_get_error_string().str;
+  rmw_reset_error();
+
+  // Destroyed a valid wait_set
+  rmw_ret_t ret = rmw_destroy_wait_set(wait_set);
+  EXPECT_EQ(ret, RMW_RET_OK) << rcutils_get_error_string().str;
+  rmw_reset_error();
+
+  // Try to create a wait_set using a invalid argument
+  wait_set = rmw_create_wait_set(nullptr, 0);
+  EXPECT_EQ(wait_set, nullptr) << rcutils_get_error_string().str;
+  rmw_reset_error();
+
+  // Battle test rmw_create_wait_set.
+  RCUTILS_FAULT_INJECTION_TEST(
+  {
+    wait_set = rmw_create_wait_set(&context, 0);
+
+    int64_t count = rcutils_fault_injection_get_count();
+    rcutils_fault_injection_set_count(RCUTILS_FAULT_INJECTION_NEVER_FAIL);
+
+    if (wait_set != nullptr) {
+      ret = rmw_destroy_wait_set(wait_set);
+      EXPECT_EQ(ret, RMW_RET_OK) << rcutils_get_error_string().str;
+    } else {
+      rmw_reset_error();
+    }
+    rcutils_fault_injection_set_count(count);
+  });
+}
+
+TEST_F(CLASSNAME(TestWaitSet, RMW_IMPLEMENTATION), rmw_wait)
+{
+  size_t number_of_subscriptions = 1;
+  size_t number_of_guard_conditions = 1;
+  size_t number_of_clients = 1;
+  size_t number_of_services = 1;
+  size_t number_of_events = 1;
+  size_t num_conditions =
+    number_of_subscriptions +
+    number_of_guard_conditions +
+    number_of_clients +
+    number_of_services +
+    number_of_events;
+
+  // Created a valid wait_set
+  rmw_wait_set_t * wait_set = rmw_create_wait_set(&context, num_conditions);
+  EXPECT_NE(wait_set, nullptr);
+  rmw_reset_error();
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    rmw_ret_t ret = rmw_destroy_wait_set(wait_set);
+    rmw_reset_error();
+    EXPECT_EQ(ret, RMW_RET_OK) << rcutils_get_error_string().str;
+  });
+
+  // Call rmw_wait with invalid arguments
+  rmw_ret_t ret = rmw_wait(nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr);
```

Comment:
@ahcorde nit: since the wait set is the only one invalid here, perhaps using named variables equal to `nullptr` for the other arguments would help readability.